psydac/linalg/topetsc.py

```diff
@@ -0,0 +1,111 @@
+import numpy as np
+
+from psydac.linalg.block import BlockVectorSpace, BlockVector, BlockMatrix
+from psydac.linalg.stencil import StencilVectorSpace, StencilVector
+from scipy.sparse import coo_matrix, bmat
+
+from mpi4py import MPI
+
+def flatten_vec( vec ):
+    """ Return the flattened 1D array values and indices corresponding to the given vector.
+
+    Parameters
+    ----------
+    vec : <Vector>
+      Psydac Vector to be flattened
+
+    Returns
+    -------
+    array : numpy.ndarray
+        A copy of the data array collapsed into one dimension.
+
+    indices: numpy.ndarray
+        The global indices the data array collapsed into one dimension.
+    """
+
+    if isinstance(vec, StencilVector):
+        npts = vec.space.npts
+        idx = tuple( slice(m*p,-m*p) for m,p in zip(vec.pads, vec.space.shifts) )
+        shape = vec._data[idx].shape
+        starts = vec.space.starts
+        indices = np.array([np.ravel_multi_index( [s+x for s,x in zip(starts, xx)], dims=npts,  order='C' ) for xx in np.ndindex(*shape)] )
+        data = vec._data[idx].flatten()
+        vec = coo_matrix(
+                    (data,(indices,indices)),
+                    shape = [np.prod(npts),np.prod(npts)],
+                    dtype = vec.space.dtype)
+
+    elif isinstance(vec, BlockVector):
+        vecs = [vec_tocoo(b) for b in vec.blocks]
```

Comment:
Does the function `vec_tocoo` still exist?

Reply:
not really, I changed its name to flatten_vec

Reply:
This is a bug then, and it should be fixed. Clearly this if-condition was not covered by the unit tests. Please make sure that it is tested instead.

Reply:
I assume that you are running the PETSc unit tests on your machine... Are you? :sweat_smile:

Reply:
I didn't in the last commit, but now it's working [2a46162](https://github.com/pyccel/psydac/pull/244/commits/2a46162c9f5bdb97d64d692f620d5aac6d0d2efc)